dotnet/CoreLib/AppBuilders/DependencyInjection.cs

```diff
@@ -120,170 +131,176 @@ public static void UseSearchClient(this IServiceCollection services, SemanticMem
+                    case Distributed:
+                        return
+                            new DistributedPipelineOrchestrator(
+                                serviceProvider.GetRequiredService<IContentStorage>(),
+                                serviceProvider.GetRequiredService<IMimeTypeDetection>(),
+                                serviceProvider.GetQueueFactory(),
+                                serviceProvider.GetRequiredService<ILogger<DistributedPipelineOrchestrator>>());
+
+                    default:
+                        throw new NotImplementedException($"Orchestration type '{config.Orchestration}' not available");
+                }
+            });
+
+        return services;
     }
 
-    public static void UseAzureBlobStorage(this IServiceCollection services, SemanticMemoryConfig config)
+    private static IContentStorage UseAzureBlobStorage(this IServiceProvider serviceProvider)
     {
         const string AzureIdentity = "AZUREIDENTITY";
         const string ConnectionString = "CONNECTIONSTRING";
 
-        // Make configuration available
-        services.AddSingleton(config.ContentStorage.AzureBlobs);
+        var config = serviceProvider.GetRequiredService<SemanticMemoryConfig>();
 
         switch (config.ContentStorage.AzureBlobs.Auth.ToUpperInvariant())
         {
             case AzureIdentity:
-                services.AddSingleton<IContentStorage>(serviceProvider => new AzureBlob(
-                    config.ContentStorage.AzureBlobs.Account,
-                    config.ContentStorage.AzureBlobs.EndpointSuffix,
-                    serviceProvider.GetService<ILogger<AzureBlob>>()));
-                break;
+                return
+                    new AzureBlob(
+                        config.ContentStorage.AzureBlobs.Account,
+                        config.ContentStorage.AzureBlobs.EndpointSuffix,
+                        serviceProvider.GetService<ILogger<AzureBlob>>());
 
             case ConnectionString:
-                services.AddSingleton<IContentStorage>(serviceProvider => new AzureBlob(
-                    config.ContentStorage.AzureBlobs.ConnectionString,
-                    config.ContentStorage.AzureBlobs.Container,
-                    serviceProvider.GetService<ILogger<AzureBlob>>()));
-                break;
+                return
+                    new AzureBlob(
+                        config.ContentStorage.AzureBlobs.ConnectionString,
+                        config.ContentStorage.AzureBlobs.Container,
+                        serviceProvider.GetService<ILogger<AzureBlob>>());
 
             default:
                 throw new NotImplementedException($"Azure Blob auth type '{config.ContentStorage.AzureBlobs.Auth}' not available");
         }
     }
 
-    public static void UseDistributedPipeline(this IServiceCollection services, SemanticMemoryConfig config)
+    private static QueueClientFactory GetQueueFactory(this IServiceProvider serviceProvider)
```

Comment:
not sure about this. I would ask to ignore the dependency injection approach because I have a lot going on here and planning to make some heavy changes during August.

Is the existing approach blocking the integration with CC?

Reply:
Yes, it does create a bit a friction with CC.  To me, the existing DI seems a little upside down.  Honestly, I think this sets the stage for better evolution.

For more complex injection patterns, introducing provider pattern can often simplify how this is managed...as opposed to creating more injection points.